File: 올바른괄호.py
def check(s) :
    stack = []
    for w in s :
        if len(stack)>0 and stack[-1] == "(" and w == ")" :
            stack.pop()
        else :
            stack.append(w)

    if len(stack) == 0 :
        return True
    else :
        return False


def solution(s):
    return check(s)

File: 코딩테스트강의/자료구조(스택,큐)/올바른 괄호.js
str = '(()(()))(()';
stack = [];
answer = 'YES';
for (let x of str) {
  if (x === '(') stack.push(x);
  if (x === ')') {
    if (stack.length) {
      stack.pop();
    } else {
      break;
    }
  }
}

if (stack.length > 0) answer = 'NO';
console.log(answer);
